crates/sui-graphql-rpc/src/types/transaction_block_kind/change_epoch.rs

```diff
@@ -0,0 +1,160 @@
+
+use async_graphql::connection::{Connection, Edge};
+use async_graphql::*;
+use move_binary_format::errors::PartialVMResult;
+use move_binary_format::CompiledModule;
+use sui_types::{
+    digests::TransactionDigest, object::Object as NativeObject,
+    transaction::ChangeEpoch as NativeChangeEpochTransaction,
+};
+
+use crate::context_data::db_data_provider::validate_cursor_pagination;
+use crate::types::sui_address::SuiAddress;
+use crate::{
+    context_data::db_data_provider::PgManager,
+    error::Error,
+    types::{
+        big_int::BigInt, date_time::DateTime, epoch::Epoch, move_package::MovePackage,
+        object::Object,
+    },
+};
+
+#[derive(Clone, PartialEq, Eq)]
+pub(crate) struct ChangeEpochTransaction(pub NativeChangeEpochTransaction);
+
+#[Object]
+impl ChangeEpochTransaction {
+    /// The next (to become) epoch.
+    async fn epoch(&self, ctx: &Context<'_>) -> Result<Epoch> {
+        ctx.data_unchecked::<PgManager>()
+            .fetch_epoch_strict(self.0.epoch)
+            .await
+            .extend()
+    }
+
+    /// The protocol version in effect in the new epoch.
+    async fn protocol_version(&self) -> u64 {
+        self.0.protocol_version.as_u64()
+    }
+
+    /// The total amount of gas charged for storage during the previous epoch (in MIST).
+    async fn storage_charge(&self) -> BigInt {
+        BigInt::from(self.0.storage_charge)
+    }
+
+    /// The total amount of gas charged for computation during the previous epoch (in MIST).
+    async fn computation_charge(&self) -> BigInt {
+        BigInt::from(self.0.computation_charge)
+    }
+
+    /// The SUI returned to transaction senders for cleaning up objects (in MIST).
+    async fn storage_rebate(&self) -> BigInt {
+        BigInt::from(self.0.storage_rebate)
+    }
+
+    /// The total gas retained from storage fees, that will not be returned by storage rebates when
+    /// the relevant objects are cleaned up (in MIST).
+    async fn non_refundable_storage_fee(&self) -> BigInt {
+        BigInt::from(self.0.non_refundable_storage_fee)
+    }
+
+    /// Time at which the next epoch will start.
+    async fn start_timestamp(&self) -> Option<DateTime> {
+        DateTime::from_ms(self.0.epoch_start_timestamp_ms as i64)
+    }
+
+    /// System packages (specifically framework and move stdlib) that are written before the new
+    /// epoch starts, to upgrade them on-chain. Validators write these packages out when running the
+    /// transaction.
+    async fn system_package_connection(
+        &self,
+        first: Option<u64>,
+        after: Option<String>,
+        last: Option<u64>,
+        before: Option<String>,
+    ) -> Result<Connection<String, MovePackage>> {
+        // TODO: make cursor opaque (currently just an offset).
+        validate_cursor_pagination(&first, &after, &last, &before).extend()?;
+
+        let total = self.0.system_packages.len();
+
+        let mut lo = if let Some(after) = after {
```

Comment:
suggestion: possible to have this as a separate function?

Reply:
Indeed!  The Cursor framework is coming up on my queue, and I'm conspicuously not abstracting this cursor code until I start on that so I can do it all in one go.

Reply:
That sounds good! We'd also need to test these well, as I am always scared of subtle bugs introduced in this if-else statements.